haskell/src/Lentille/GraphQL.hs

```diff
@@ -99,71 +104,77 @@ instance From PageInfo Text where
       <> maybe "" show totalCount
       <> (if hasNextPage then " (has next page)" else "")
 
-data RateLimit = RateLimit {used :: Int, remaining :: Int, resetAt :: Text}
+data RateLimit = RateLimit {used :: Int, remaining :: Int, resetAt :: UTCTime}
   deriving (Show)
 
 instance From RateLimit Text where
-  from RateLimit {..} = "remains:" <> show remaining <> ", reset at: " <> resetAt
+  from RateLimit {..} = "remains:" <> show remaining <> ", reset at: " <> show resetAt
 
 streamFetch ::
   (MonadGraphQLE m, Fetch a, FromJSON a, Show a) =>
   GraphClient ->
   -- | query Args constructor, the function takes a cursor
   (Maybe Text -> Args a) ->
+  -- | an action to get a RateLimit record
+  (Maybe (GraphClient -> m RateLimit)) ->
   -- | query result adapter
   (a -> (PageInfo, Maybe RateLimit, [Text], [b])) ->
   Stream (Of b) m ()
-streamFetch client@GraphClient {..} mkArgs transformResponse = go Nothing
+streamFetch client@GraphClient {..} mkArgs getRateLimitM transformResponse = go Nothing
   where
     log :: MonadLog m => Text -> m ()
     log = mLog . Log Macroscope . LogGraphQL lc
       where
         lc = LogCrawlerContext index crawler
 
-    request pageInfoM waitUntilM = do
-      case waitUntilM of
-        Just waitUntil -> do
-          log $ "Reached Quota limit. Waiting until reset date: " <> show waitUntil
-          holdOnUntil waitUntil
-        Nothing -> pure ()
+    holdOnIfNeeded :: (MonadTime m, MonadLog m) => Maybe RateLimit -> m ()
+    holdOnIfNeeded = \case
+      Nothing -> pure ()
+      Just rl -> do
+        if remaining rl <= 0
+          then do
+            let resetAtTime = resetAt rl
+            log $ "Reached Quota limit. Waiting until reset date: " <> show resetAtTime
+            holdOnUntil resetAtTime
+          else pure ()
+        pure ()
+      where
+        holdOnUntil :: (MonadTime m) => UTCTime -> m ()
+        holdOnUntil resetTime = do
+          currentTime <- mGetCurrentTime
+          let delaySec = diffUTCTimeToSec resetTime currentTime + 1
+          mThreadDelay $ delaySec * 1_000_000
+          where
+            diffUTCTimeToSec a b =
+              truncate (realToFrac . nominalDiffTimeToSeconds $ diffUTCTime a b :: Double) :: Int
```

Comment:
That seems quite useful, perhaps this can be moved to the `Monocle.Class` module?